### ISSUE TITLE
potential bug in shape layer edge

### ISSUE THREAD

**Issue by yunhal** [napari/napari#5770]
## 🐛 Bug

Napari's shape layer edge is correctly visualized when feeding [**[20 20]**  [50 70]  [80 20] ], but incorrectly shown when feeding [**[20 20]** [50 70] [80 20] **[20 20]** ].  You will get the array like the latter, when extracting coordinates from shapely polygon, e.g., polygon.exterior.coords.  

## To Reproduce

```
from shapely.geometry import Polygon
import napari
import numpy as np

# Create Polygon
exterior = [(20, 20), (50, 70), (80, 20)]
interior = [(40, 25), (40, 45), (60, 45), (60, 25)]
poly = Polygon(exterior, holes=[interior])

print("np array from poly exterior or interior has one extra length")
print("exterior", np.array(exterior))
print("poly exterior", np.array(list(poly.exterior.coords)))
print("poly interior", np.array(poly.interiors[0]))

viewer = napari.Viewer()
layer = viewer.add_shapes(np.array(exterior), name = "exterior", shape_type=['polygon'])
layer = viewer.add_shapes(np.array(interior), name = "interior", shape_type=['polygon'])

viewer = napari.Viewer()
layer = viewer.add_shapes(np.array(list(poly.exterior.coords)), name = "poly exterior", shape_type=['polygon'])
layer = viewer.add_shapes(np.array(poly.interiors[0]), name = "poly interior", shape_type=['polygon'])

```

Please see the pink circle for the difference:
<img width="1631" alt="Screen Shot 2023-04-25 at 10 35 26 AM" src="https://user-images.githubusercontent.com/28872346/234321729-4d80c9dd-f52c-4fdd-9bb1-bf6482d2f6bc.png">

## Environment

napari: 0.4.16
Platform: macOS-12.6-x86_64-i386-64bit
System: MacOS 12.6
Python: 3.10.6 | packaged by conda-forge | (main, Aug 22 2022, 20:43:44) [Clang 13.0.1 ]
Qt: 5.15.4
PyQt5: 5.15.7
NumPy: 1.23.3
SciPy: 1.9.1
Dask: 2022.9.2
VisPy: 0.10.0

OpenGL:
- GL version: 2.1 ATI-4.8.101
- MAX_TEXTURE_SIZE: 16384

Screens:
- screen 1: resolution 1920x1080, scale 2.0

Plugins:
- brainglobe-napari-io: 0.1.5
- brainreg-napari: 0.0.5rc0
- brainreg-segment: 0.2.16
- console: 0.0.6
- napari-animation: 0.0.5
- napari-dataset: 0.1.dev81+g454e834
- napari-geojson: 0.1.3
- napari-hdf5-netcdf-viewer: 0.0.1
- napari-svg: 0.1.6
- ndtiffs: 0.1.2
- scikit-image: 0.4.16
 - Any other relevant information:

## Additional context

This is related to my recent polygon bug report: https://github.com/napari/napari/issues/5673.

**Comment by yunhal** [comment]
Visualizing a circle shows the edge bug more clearly. 

![image](https://user-images.githubusercontent.com/28872346/234361384-25e32fff-21b8-4167-bc71-5e4bbe8b9aa8.png)

**Comment by psobolewskiPhD** [comment]
I can reproduce the issue with the corner of the triangle on main, but I also get a traceback with the second polygon:
```
TypeError                                 Traceback (most recent call last)
Cell In[1], line 21
     19 viewer = napari.Viewer()
     20 layer = viewer.add_shapes(np.array(list(poly.exterior.coords)), name = "poly exterior", shape_type=['polygon'])
---> 21 layer = viewer.add_shapes(np.array(poly.interiors[0]), name = "poly interior", shape_type=['polygon'])

File ~/Dev/napari/napari/components/viewer_model.py:4, in add_shapes(self, data, ndim, features, feature_defaults, properties, property_choices, text, shape_type, edge_width, edge_color, edge_color_cycle, edge_colormap, edge_contrast_limits, face_color, face_color_cycle, face_colormap, face_contrast_limits, z_index, name, metadata, scale, translate, rotate, shear, affine, opacity, blending, visible, cache, experimental_clipping_planes)
      1 from __future__ import annotations
      3 import inspect
----> 4 import itertools
      5 import os
      6 import warnings

File ~/Dev/napari/napari/layers/shapes/shapes.py:421, in Shapes.__init__(self, data, ndim, features, feature_defaults, properties, property_choices, text, shape_type, edge_width, edge_color, edge_color_cycle, edge_colormap, edge_contrast_limits, face_color, face_color_cycle, face_colormap, face_contrast_limits, z_index, name, metadata, scale, translate, rotate, shear, affine, opacity, blending, visible, cache, experimental_clipping_planes)
    387 def __init__(
    388     self,
    389     data=None,
   (...)
    419     experimental_clipping_planes=None,
    420 ) -> None:
--> 421     if data is None or len(data) == 0:
    422         if ndim is None:
    423             ndim = 2

TypeError: len() of unsized object
```

**Comment by yunhal** [comment]
Hi Peter, Thanks for taking a look at this issue.  I tried again the code sampled above and was able to execute without any error.  I don't understand why you get an error.  It looks like the error comes from "poly.interiors[0]" in your case, which appears empty??  The "np.array(poly.interiors[0])" should be [ [40. 25.]. [40. 45.]. [60. 45.]  [60. 25.] [40. 25.] ], which is almost same as "interior" array except having one additional index.

**Comment by psobolewskiPhD** [comment]
We're both on macOS but with different envs, so maybe it's related to napari or dependency versions.

**Comment by yunhal** [comment]
The error that Peter encountered is related to the new shapely version - I got the same error after updating 1.8.5 to 2.0.1.  

Replacing the last line of the code on the top of this page with the one below will resolve the problem. 

```
import shapely
layer = viewer.add_shapes(shapely.get_coordinates(poly.interiors[0]), name = "poly interior", shape_type=['polygon'])
```

**Comment by psobolewskiPhD** [comment]
Yup, I can confirm. The code from above from @yunhal resolves the TypeError but the clipped corners remain.

**Comment by psobolewskiPhD** [comment]
Note this also affects geojson-derived shapes which use LinearRings for polygons. See: https://napari.zulipchat.com/#narrow/channel/212875-general/topic/napari.20Shapes.2C.20geojson.20.28and.20QuPath.29/with/518121551

**The visual element of this is resolved in 0.6.0, due to the changes in the triangulation pipeline that deal with duplicate vertexes.** 
However, the duplicate vertex **is present and can be interacted with** which is unexpected and changes the triangle into a trapezoid.

```
import shapely
from shapely.geometry import Polygon
import napari
import numpy as np

# Create Polygon
exterior = [(20, 20), (50, 70), (80, 20)]
interior = [(40, 25), (40, 45), (60, 45), (60, 25)]
poly = Polygon(exterior, holes=[interior])

print("np array from poly exterior or interior has one extra length")
print("exterior", np.array(exterior))
print("poly exterior", np.array(list(poly.exterior.coords)))
print("poly interior", np.array(poly.interiors[0]))

viewer = napari.Viewer(title = "non-RING")
layer = viewer.add_shapes(np.array(exterior), name = "exterior", shape_type=['polygon'])
laye2r = viewer.add_shapes(np.array(interior), name = "interior", shape_type=['polygon'])

viewer = napari.Viewer(title = "RING")
layer = viewer.add_shapes(np.array(list(poly.exterior.coords)), name = "poly exterior", shape_type=['polygon'])
layer2 = viewer.add_shapes(shapely.get_coordinates(poly.interiors[0]), name = "poly interior", shape_type=['polygon'])

```

So I'm of two minds about closing this. Before there was a visual indicator of 2 vertexes. That's been "fixed" but really just swallowed. If napari is going to remove duplicate vertices when rendering maybe they should also not be present in the polygon? In other words use rings to indicate polygons with the implication that duplicate vertexes mark start and end? In fact after #6654 napari supports passing rings to indicate polygons with holes, but even in the example from that PR this shows up:

<img width="1331" alt="Image" src="https://github.com/user-attachments/assets/6eaf519e-3a10-4a77-8562-fbbc0fe9b0bd" />

**Comment by jni** [comment]
@psobolewskiPhD duplicate vertices in the source data can't *only* indicate closed polygons, because you need them to indicate the holes, as part of a single data sequence.

We can leave this open as a memento but I might suggest that we close and instead open a new issue about our polygon representations based on [my message on the Zulip thread](https://napari.zulipchat.com/#narrow/channel/212875-general/topic/napari.20Shapes.2C.20geojson.20.28and.20QuPath.29/near/518766553):

> Well, we actually allow rings anywhere, and then convert them to non-rings + a boolean flag for later processing:
> 
> https://github.com/napari/napari/blob/c6c71329431b942a8e9379acc7cebda741640465/napari/layers/shapes/_accelerated_triangulate_python.py#L377
> 
> Which is actually important e.g. for the UI because you don't want to be able to grab the duplicated vertex and mess everything up, as you showed @**Peter Sobolewski**. So if anything your example suggests we should use rings *less*. :joy: 
> 
> It's very straightforward to convert closed shapes to rings and vice versa at IO time, so I'm not sure how critical it is to harmonise our representations everywhere. The *big* things to get out from this though (imho) are:
> 
> - we need different representations for different uses, and we should be more strict about which representation we use in which part of the code, and have single mechanisms for converting from each format to another:
>    - a "canonical" data representation, which goes in Layer.data, which inputs get converted to and from which we make outputs.
>    - a *different* representation for the data that gets drawn on screen — duplicate vertices are forbidden by triangulation algorithms.
>    - a *third* representation for the data that gets *manipulated* by edits, as shown by the Lesotho Duplicate Vertex Scandal of 2025. 😜 (btw there is a corresponding double vertex in SA, @**Peter Sobolewski**! :scream:) (I'm not 100% sure this needs to be different from the second representation, but I suspect it does, for one thing, in order to be able to draw *new* holes.)
> - Once we are clear on which representation is used where, it should straightforward to convert to/from GeoJSON or any other file format at IO time. I don't think it's absolutely necessary to exactly match the GeoJSON format in-memory, but if it's easy and doesn't have performance implications, we might as well.

But we can keep discussing until we agree on the core of the issue, I'm not 100% on the above. 😅 And then open the issue with a clear path forward.

**Comment by psobolewskiPhD** [comment]
@jni I think I'm leaning to marking this closed -- the visual issue is solved -- and opening a new issue for duplicated vertexes if passed a ring. That can be either solved by using the de-duplicated array ala https://github.com/napari/napari/blob/c6c71329431b942a8e9379acc7cebda741640465/napari/layers/shapes/_accelerated_triangulate_python.py#L377 or by changing out internal representation to be more strict, as you suggest in the zulpi thread.

**Comment by Czaki** [comment]
We should not change that before we move Shapes drawing into overlays.

**Comment by psobolewskiPhD** [comment]
Ok. Shall we close this and open a new issue for duplicate vertexes when displaying polygons with holes?
And I will PR napari-geojson for plain polygons to close ring?

**Comment by psobolewskiPhD** [comment]
Closing to focus discussion in https://github.com/napari/napari/issues/7941
